Read json from the bagit package.

MancunianSam opened the pull request:
We've got rid of the three CSV files and replaced it with a single bagit
json. This change reads from that file and writes it to Dynamo.

It's more or less a copy of the json in the bagit package into Dynamo.
We're calculating the parent path from the ID and adding in the
checksums from the bag metadata and the file extensions and adding them
to the dynamo json.

This is all done with json objects and string keys. It's a bit fiddly
but as there'll be fields that we won't know in advance, we can't use a
case class.

techncl reviewed (commented):
Looks good! Just some additional comments. Also,

1. In the DiscoveryServiceTest, in the `getAssetFromDiscoveryApi` method, there is a possibility of receiving a "Cannot find asset with citable reference $citableReference" error; should we test for this?
2. just noticed, in your previous PR where you replaced the HTML codes, you added a test for it in the LambdaTest file but not in the  DiscoveryServiceTest, which means that if you change the LambdaTest code, there will no longer be a test for it; should we add a test for it here?
3. In the same file, there is no test for the`titleWithoutBackslashes` code to ensure that backslashes will be removed

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -62,19 +62,30 @@ class DiscoveryService(discoveryBaseUrl: String, backend: SttpBackend[IO, Fs2Str
       val departmentTableEntry = departmentDiscoveryAsset
         .map(tableEntry)
-        .getOrElse(DynamoTable(input.batchId, randomUuidGenerator(), "", "Unknown", folder, "", ""))
+        .getOrElse(Map("batchId" -> Str(input.batchId), "id" -> Str(randomUuidGenerator().toString), "name" -> Str("Unknown"), "type" -> Str(ArchiveFolder.toString)))
```
The `Unknown` folder is an archive folder so we should only ever have a single Unknown folder. The upsert lambda should recognise if there's already an Unknown top level folder and not create a new one if there already is.

MancunianSam commented:
There is a test for "Cannot find asset with citeable reference" https://github.com/nationalarchives/dr2-ingest-mapper/blob/DR2-1209-read-flat-json-from-bagit/src/test/scala/uk/gov/nationalarchives/DiscoveryServiceTest.scala#L98

I've added in the html codes into the test data for DiscoveryServiceTest. 
I've added some backslashes into the test data as well to cover that.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -62,19 +62,30 @@ class DiscoveryService(discoveryBaseUrl: String, backend: SttpBackend[IO, Fs2Str
         .map(tableEntry)
-        .map(_.copy(parentPath = departmentTableEntry.id.toString))
-      DepartmentAndSeriesTableData(departmentTableEntry, seriesTableEntryOpt)
+        .map(jsonMap => jsonMap ++ Map("parentPath" -> departmentTableEntry("id")))
```
Do you mean the `departmentTableEntry` variable should be `departmentTableEntryMap`? That sounds sensible. The other `tableEntry` is a function which you've put a comment about below.

techncl replied:
Sorry, yes ,`departmentTableEntryMap`

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/Lambda.scala:
```diff
@@ -59,23 +64,24 @@ class Lambda extends RequestStreamHandler {
+        .mapValues(_.map(jsonObj => UUID.fromString(jsonObj("id").str)))
+        .toMap
 
       val stateData = StateData(
```
This is the state that's written to the output stream so `StateOutput` makes sense. I've changed it to that.

techncl replied:
Ah yes, I was thinking "input" to the next thing in the pipeline but yeah, it makes sense from the perspective of this Lambda, to be "output".

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -1,26 +1,94 @@
+            bagitManifestString
+              .split("\n")
+              .map { row =>
+                val eachColumn = row.split(" ")
```
`rowAsArray` is good.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -1,26 +1,94 @@
+    parseFileFromS3(
+      input,
+      "manifest-sha256.txt",
+      file => {
```
I can get rid of the whole line and just to `_.` I've done that.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -1,26 +1,94 @@
+  private def parseFileFromS3[T](input: Input, name: String, decoderPipe: Pipe[IO, String, T]): IO[List[T]] = {
     for {
       pub <- s3.download(input.s3Bucket, s"${input.s3Prefix}$name")
       csvString <- pub
```
No, I've changed it to `s3FileString`

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -32,83 +100,37 @@ class MetadataService(s3: DAS3Client[IO]) {
-
-  case class FileMetadata(identifier: UUID, parentPath: String, name: String, fileSize: Long, title: String) extends Metadata
-
   case class BagitManifest(checksum: String, filePath: String)
```
`BagitManifestRow` is good.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -1,26 +1,94 @@
+    }.toMap
+
+    def searchParentIds(parentIdOpt: Option[UUID]): String = {
+      val nextParent = parentIdOpt.flatMap(idToParentId.get).flatten
```
I've swapped them and renamed nextParent.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -1,26 +1,94 @@
+                    .map(Str)
+                    .getOrElse(Null)
+                else Null
+              val objectMap =
```
`metadataMap` makes sense.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/Lambda.scala:
```diff
@@ -40,17 +35,27 @@ class Lambda extends RequestStreamHandler {
-    typeObject => typeObject.toString
-  )
+  implicit val dynamoTableFormat: Typeclass[Obj] = new Typeclass[Obj] {
+    override def read(av: DynamoValue): Either[DynamoReadError, Obj] = {
```
That was auto-generated. I don't know what it stands for either. I've changed it to `dynamoValue`

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/Lambda.scala:
```diff
@@ -40,17 +35,27 @@ class Lambda extends RequestStreamHandler {
+    }
 
-  implicit val dynamoTableFormat: Typeclass[DynamoTable] = deriveDynamoFormat[DynamoTable]
+    override def write(t: Obj): DynamoValue = {
```
I've changed this to `jsonObject`

MancunianSam commented on src/test/scala/uk/gov/nationalarchives/MetadataServiceTest.scala:
```diff
@@ -1,128 +1,81 @@
+  s"parseBagManifest" should "return the correct row values" in {
+    val input = Input("testBatch", "bucket", "prefix/", Option("T"), Option("T TEST"))
+    val id = UUID.randomUUID()
+    val validCsv = s"checksum $id"
```
Ok. it was a generic csv before and I didn't update the name.

MancunianSam commented on src/test/scala/uk/gov/nationalarchives/MetadataServiceTest.scala:
```diff
@@ -1,128 +1,81 @@
+  s"parseBagManifest" should "return the same response for both fields if there is only one column" in {
+    val input = Input("testBatch", "bucket", "prefix/", Option("T"), Option("T TEST"))
+    val id = UUID.randomUUID()
+    val invalidCsv = s"invalid"
```
I've changed it to `val invalidBagitManifestRow = "onlyOneColumn"`
I've also changed the `parseBagManifest` method to check if there are two columns and raise an error if there are and updated the test to expect the error.

MancunianSam commented on src/test/scala/uk/gov/nationalarchives/MetadataServiceTest.scala:
```diff
@@ -132,10 +85,19 @@ class MetadataServiceTest extends AnyFlatSpec with MockitoSugar with TableDriven
   )
   forAll(departmentSeriesTable) { (departmentId, seriesIdOpt) =>
     "metadataToDynamoTables" should s"return a list of tables with the correct prefix for department $departmentId and series ${seriesIdOpt.getOrElse("None")}" in {
```
I've updated the test name. 
The metadata json is just a test example of the real data we'll be getting through for judgments so I think it makes sense to have it all in one place in the test. I think if I try to abstract it out, I'd need a parameter in the table for each attribute in each json object which might be a bit unweildy. Or do you mean something else?

MancunianSam commented on src/test/scala/uk/gov/nationalarchives/MetadataServiceTest.scala:
```diff
@@ -145,57 +107,57 @@ class MetadataServiceTest extends AnyFlatSpec with MockitoSugar with TableDriven
       )
-      checkTableRows(result, List(folderId), 1, DynamoTable(batchId, folderId, s"$prefix", folderMetadata.head.name, Folder, folderMetadata.head.title, ""))
-      checkTableRows(result, List(assetId), 1, DynamoTable(batchId, assetId, s"$prefix/$folderId", assetMetadata.head.title, Asset, "", ""))
+      checkTableRows(result, List(folderId), DynamoTable(batchId, folderId, s"$prefix", "TestName", ArchiveFolder, "TestTitle", ""))
```
Good spot, copy and paste error.

techncl reviewed (approved):
Thanks.
